bluetooth: BLE GATT device information service

lemrey opened the pull request:
This pull request moves the device information service to a new location under /subsys/bluetooth and adds a Kconfig entry to enable and configure the service. When enabled, the service registers itself automatically with the BLE stack.

FYI: @carlescufi, @joerchan.

carlescufi commented:
@rakons can you take a look to see if this fits your requirements?

rakons commented on drivers/sensor/mockup/battery/battery.c:
```diff
@@ -0,0 +1,105 @@
+	return 0;
+}
+
+static int batter_channel_get(struct device *dev, enum sensor_channel chan,
```
spelling - batter**y**

rakons commented on include/bluetooth/services/bas.h:
```diff
@@ -12,8 +13,7 @@
 
-void bas_init(void);
-void bas_notify(void);
+int bas_notify(void);
```
Is it an API function? Then why there is no comment at all?

lemrey replied:
Because there wasn't any to begin with and the PR is focused on something else.

jhedberg replied:
There should then at least be a follow-up patch (either in this PR, or another one after it) that prefixes these appropriately. Everything under include/bluetooth must at least have a bt_* prefix. I'll let you add more namespacing if you feel like it's appropriate (e.g. bt_gatt_*)

lemrey replied:
@jhedberg I have prefixed it with `bt_gatt_` and added a small doxy section.

rakons commented on subsys/bluetooth/services/bas.c:
```diff
@@ -0,0 +1,140 @@
+	if (trigger->type == SENSOR_TRIG_DATA_READY &&
+	    trigger->chan == SENSOR_CHAN_SOC) {
+		/* fetch battery level and send notifications, if enabled */
+		charge_lvl_fetch();
```
What is the reason of complicating the data flow so much? Cannot the sensor just provide the data here as a parameter to this callback?

lemrey replied:
I don't see how, with the current sensor.h interface.

lemrey commented:
Hi @rakons , this PR is about the Device Information service. It contains commits from PR https://github.com/zephyrproject-rtos/zephyr/pull/9735 where I made changes to the battery service and where I branched from. If you have comments on the battery service could you please write them in that pull request? Thanks :)

rakons commented on samples/bluetooth/peripheral_csc/prj.conf:
```diff
@@ -2,5 +2,8 @@ CONFIG_ENTROPY_GENERATOR=y
 CONFIG_BT_PERIPHERAL=y
+CONFIG_BT_GATT_BAS=y
+CONFIG_BT_GATT_DIS=y
+CONFIG_BT_GATT_DIS_MANUF="ACME"
```
Yea... good joke. Be careful with that as such company really exist, even more than one:
https://www.acme.eu/en-us
https://acme.com/
http://acme-world.com/about/

rakons replied:
In USB library we have: "USB manufacturer name". Looks just more professional.

rakons reviewed (changes requested):
Basically... my felling is that this service is overcomplicated. I am a big fan of KISS strategy - from the BAS I would expect that it realizes the BAS itself and nothing more - having just bas_notify(u8_t level) should be fairly enough for sensor feedback function. The whole sensor connection and logic may be realized outside then. Forcing the user to use sensor API complicates the whole thing and I cannot see real gain here. The example code may look simpler maybe - but the cost is in overcomplicated mechanics at the bottom. You have to write much more code to support your own sensor than just by using simple notify function.

lemrey commented on drivers/sensor/mockup/battery/battery.c:
```diff
@@ -0,0 +1,105 @@
+
+	data->battery_lvl = data->battery_lvl > CONFIG_BATTERY_SENSOR_LVL_MIN
+				? data->battery_lvl - 1
+				: CONFIG_BATTERY_SENSOR_LVL_MAX;
```
This is a mockup that simulates a decrease in the battery level, when the minimum level is reached, the battery level resets. This maintains the behavior of the previous BAS implementation.

rakons replied:
It would be nice to see some comment like:
/* Simulating a decrease in the battery level. */

This makes it obvious. Or don't. That may be only me who feels a little lost here as I am still not totally in the architecture and code structure.

Vudentz replied:
Id change the lvl on the work function not on fetch otherwise if someone fetch it again t would change the level without signalling which imo is not correct since it would not honor CONFIG_BATTERY_SENSOR_TRIG_PERIOD.

lemrey commented:
> why it contains BAS commits too?

It depends on the CMakeFile and Kconfig files under /subsys/bluetooth that were added in the other PR.

JoeHut commented on subsys/bluetooth/services/Kconfig.dis:
```diff
@@ -0,0 +1,109 @@
+config BT_GATT_DIS_LOG_LEVEL_INF
+	bool "Info"
+config BT_GATT_DIS_LOG_LEVEL_DBG
+	bool "Debug"
```
This is inconsistent with how it is done usually right?

jhedberg replied:
True. AFAIK there's usually just a single integer option for the level with `range 0 4`

JoeHut commented on subsys/bluetooth/services/Kconfig.dis:
```diff
@@ -0,0 +1,109 @@
+
+ endchoice
+
+ config BT_GATT_DIS_LOG_LEVEL
```
There is a space before a couple of configs...

JoeHut commented on subsys/bluetooth/services/Kconfig.dis:
```diff
@@ -0,0 +1,109 @@
+
+config BT_GATT_DIS_MODEL
+	string "Model name"
+	default SOC
```
Is this really resolved?

rakons reviewed (changes requested):
Add Radoslaw Koppel <radoslaw.koppel@nordicsemi.no> to "Signed-off-by" :)

Vudentz commented on subsys/bluetooth/services/Kconfig.dis:
```diff
@@ -0,0 +1,109 @@
+config BT_GATT_DIS_PNP_VID
+	hex "Vendor ID"
+	range 0 0xFFFF
+	default 0
```
I wonder if we wouldn't happen to have a default as zephyr, afaik we did have something for USB which perhaps we can use here as well.

lemrey replied:
Don't know myself, I guess it could be changed later?

Vudentz commented on subsys/bluetooth/services/Kconfig.dis:
```diff
@@ -0,0 +1,109 @@
+config BT_GATT_DIS_PNP_PID
+	hex "Product ID"
+	range 0 0xFFFF
+	default 0
```
Ditto.

Vudentz commented on subsys/bluetooth/services/Kconfig.dis:
```diff
@@ -0,0 +1,109 @@
+	 The device manufacturer inside Device Information Service.
+
+config BT_GATT_DIS_PNP
+	bool "Enable PnP_ID characteristic"
```
I would probably default this to y since usually PID and VID is used to probe a driver, at least in Linux it works that way. That said if we have good defaults maybe we should remove this and always have the PNP char declared.

rakons replied:
PnP is optional characteristic. From the top of my mind it seems that it is mandatory only for HOG. If we think about DIS globally t should not be enabled by default.

Vudentz replied:
But the other information DIS carries can easily be mapped to device name, e.g: manufacturer and model, while PNP is only provided by DIS which is probably why HoG mandates it. So if you are going to use DIS Id assume it is for enabling the use of PID/VID, not to access manufacturer name, etc.

kapi-no replied:
I guess we can have PNP enabled by default, because it is still optional charactertistic for HRP (especially when it improves interoperability). However, I would still leave it configurable and possibly make all DIS characteristics configurable

joerchan replied:
I think we should make all attributes configurable then, and default to off. Other services can then select the characteristics it mandates.

lemrey replied:
Off by default.

Vudentz commented on subsys/bluetooth/services/dis.c:
```diff
@@ -0,0 +1,92 @@
+static const char dis_manuf[] = CONFIG_BT_GATT_DIS_MANUF;
+
+#if CONFIG_BT_GATT_DIS_PNP
+static const u8_t dis_pnp_id[DIS_PNP_ID_LEN] = {
```
Can't we have a struct mapping the individual field properly? That can be packed to guarantee no  padding is added.

Vudentz commented on subsys/bluetooth/services/dis.c:
```diff
@@ -0,0 +1,92 @@
+
+#define DIS_PNP_ID_LEN	7	/** Length of PnP ID Characteristic Value */
+
+static const char dis_model[] = CONFIG_BT_GATT_DIS_MODEL;
```
You can pass the config options when declaring the characteristic, then access it with attr->user_data, that way we avoid having static pointers.

lemrey replied:
Done

Vudentz commented on subsys/bluetooth/services/bas.c:
```diff
@@ -1,73 +1,129 @@
+		return -EINVAL;
+	}
+
+	u8_t lvl8 = (u8_t)(charge_lvl.val1);
```
Id have the declaration in the beginning of the function, some c standard actually forbids declaration not in the beginning of the function.

Vudentz commented on drivers/sensor/mockup/battery/battery.c:
```diff
@@ -0,0 +1,105 @@
+
+static void timer_expire(struct k_timer *dummy)
+{
+	k_work_submit(&timer_work);
```
You could have used a delayed_work here of a timer since you end up submitting a work anyway.

Vudentz commented on drivers/sensor/mockup/battery/battery.c:
```diff
@@ -0,0 +1,105 @@
+		return -ENOTSUP;
+	}
+
+	struct battery_data *data = dev->driver_data;
```
Declaration in the beginning of the function.

jhedberg commented on subsys/bluetooth/services/Kconfig.dis:
```diff
@@ -0,0 +1,109 @@
+	bool "Enable BLE GATT Device Information service"
+	default n
+
+if (BT_GATT_DIS)
```
the parenthesis are not needed (and inconsistent - I haven't seen them used like this elsewhere in the tree)

lemrey replied:
Fixed

jhedberg commented on subsys/bluetooth/services/Kconfig:
```diff
@@ -0,0 +1,17 @@
+
+endmenu
+
+endif #BT_PERIPHERAL
```
This doesn't match the actual contition, should be `BT_CONN` (I know this is because it used to be different in an earlier revision)

lemrey replied:
Fixed

jhedberg commented:
Any update on this one? Seems it needs rebasing. There are also some unanswered comments.

carlescufi commented:
> Any update on this one? Seems it needs rebasing. There are also some unanswered comments.

I don't see merge conflicts? that said @lemrey can you take a look at the unadressed comments?

lemrey committed:
subsys: bluetooth: services: dev. information service enhancements

This commit moves the BLE GATT Device Information service
from /samples/bluetooth/gatt to /subsys/bluetooth/services and adds
a Kconfig entry to enable and configure the service;
when enabled, it will register itself automatically.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>
Signed-off-by: Radoslaw Koppel <radoslaw.koppel@nordicsemi.no>

lemrey commented:
Since the other pull request (battery service) is blocked, I reworked the commits to include all the necessary dependencies for this patch in one commit. All issues pertaining DIS are addressed.

lemrey commented on subsys/bluetooth/services/Kconfig.dis:
```diff
@@ -0,0 +1,109 @@
+	range 1 2
+	default 1
+	help
+	 The Vendor ID Source field designates which organization assigned the value used in the Vendor ID field value.
```
Fixed

carlescufi commented:
@Vudentz would you mind taking another look at this one?

Vudentz reviewed (approved):
LGTM
